server: Add Authentication Service

mrtracy opened the pull request:
Adds new authentication service to the server, as outlined in the web
sessions RFC.

+ Service has two methods, UserLogin and UserLogout. Only UserLogin is
implemented.
+ UserLogin, when successful, creates an web session and sets a cookie
on the outgoing HTTP response containing the session id and secret.

Commit:
server: Add Authentication Service

Adds new authentication service to the server, as outlined in the web
sessions RFC.

+ Service has two methods, UserLogin and UserLogout. Only UserLogin is
implemented.
+ UserLogin, when successful, creates an web session and sets a cookie
on the outgoing HTTP response containing the session id and secret.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17343)
<!-- Reviewable:end -->

petermattis commented:
Heh, I can't assign a reviewer, presumably due to `embedded.go`.

@couchand and @a-robinson, can you give this a review.

couchand commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> except for the one test assertion error.  Left lots of minor style comments, feel free to take them or leave them.  Great work!

---

Reviewed 10 of 16 files at r1.
Review status: 8 of 16 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/server/api_error.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTi4m8-sL4yUNEgOJK:-KqTi4m8-sL4yUNEgOJL:b-7rhqg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/api_error.go#L15)):*
> ```Go
> // permissions and limitations under the License.
> //
> // Author: Matt Tracy (matt@cockroachlabs.com)
> ```

FYI since #17346 Author comments are no bueno.

---

*[pkg/server/authentication.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTkVfeDJLuriD0eWEA:-KqTkVfeDJLuriD0eWEB:b-psodxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L59)):*
> ```Go
> var webSessionTimeout = settings.RegisterNonNegativeDurationSetting(
> 	"server.web_session_timeout",
> 	"the duration that a newly created web session login will be valid.",
> ```

Nits:
- Looks like these descriptions don't usually have trailing punctuation.
- I'd remove "login".

---

*[pkg/server/authentication.go, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTm6nPxT9WfSBZaGq0:-KqTm6nPxT9WfSBZaGq1:bw9iaff) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L65)):*
> ```Go
> 	server     *Server
> 	executor   sql.InternalExecutor
> 	memMonitor mon.MemoryMonitor
> ```

It doesn't look like this is used anywhere?

---

*[pkg/server/authentication.go, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTn5UM7wlELZ0ERo7R:-KqTn5UM7wlELZ0ERo7S:bq4bs09) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L103)):*
> ```Go
> // session. It checks the provided credentials against the system.users table,
> // and if successful creates a new authentication session. The session's ID
> // and secret
> ```

Looks like this comment is unfinished?

---

*[pkg/server/authentication.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTpbm-vDzPpPvxvDpB:-KqTpbm-vDzPpPvxvDpC:b1hcwgt) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L163)):*
> ```Go
> 
> // verifyPassword verifies the passed username/password pair against the
> // system.users table.
> ```

To make everything super clear, it might be useful to repeat the comment you added to `CompareHashAndPassword`, specifically "If they are not equivalent, returns an error."

---

*[pkg/server/authentication.go, line 220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqU5KRYaBHcx_n22Pi9:-KqU5KRYaBHcx_n22PiA:b-cq76oj) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L220)):*
> ```Go
> 		id = int64(*datum[0].(*parser.DInt))
> 		return nil
> 	}); err != nil {
> ```

Maybe it's just me, but I find it really hard to read a multi-line `if` initialization.

---

*[pkg/server/authentication.go, line 229 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqU5jqLZRMMIcJJTp98:-KqU5jqLZRMMIcJJTp99:bq4brzs) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L229)):*
> ```Go
> // authenticationHeaderMatcher is a GRPC header matcher function, which provides
> // a conversion from GRPC headers to HTTP headers. This function is needed to
> // allow the "Set-Cookie" header to
> ```

Looks like this comment is unfinished.

---

*[pkg/server/authentication_test.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUApxBVlPerZO-4Z4m:-KqUApxBVlPerZO-4Z4n:b-6vgvdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L185)):*
> ```Go
> 		shouldAuthenticate bool
> 	}{
> 		{"azure_diamond", "hunter2", true},
> ```

I'd move these test cases out above the `for` statement.

---

*[pkg/server/authentication_test.go, line 191 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUNBAROXQOthArSHWv:-KqUNBASGawoMPbA-Bit:b-3hauxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L191)):*
> ```Go
> 		{"azure_diamond", "hunter2345", false},
> 		{"azure_diamond", "12345", false},
> 		{"azure_diamond", "*******", false},
> ```

Just for good measure, how about a suffix match like `shunter2`?

---

*[pkg/server/authentication_test.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUOOmB5zrM3QUIUvx-:-KqUOOmB5zrM3QUIUvx0:b-yhnyhy) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L270)):*
> ```Go
> 	hashedSecret := hasher.Sum(origSecret)
> 	if !bytes.Equal(sessHashedSecret, hashedSecret) {
> 		t.Fatalf("hashed value of %v was %v, wanted %v", origSecret, hashedSecret, sessHashedSecret)
> ```

Should this error also say the word "secret" somewhere?  Also, spitting out the raw bytes might not be a good idea (how does `%v` for `[]byte` look?)

---

*[pkg/server/authentication_test.go, line 274 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUOoHx37UVHlSrPMyr:-KqUOoHx37UVHlSrPMys:btghh5t) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L274)):*
> ```Go
> 
> 	// Username.
> 	if a, e := sessUsername, username; a != e {
> ```

Why the aliasing?

---

*[pkg/server/authentication_test.go, line 283 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUP6mjtiVO7JO_KhbG:-KqUP6mjtiVO7JO_KhbH:b-krnuxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L283)):*
> ```Go
> 			return fmt.Errorf("time %s was before early bound %s", actual, early)
> 		}
> 		if late.Before(early) {
> ```

:s/early/actual/

---

*[pkg/server/authentication_test.go, line 327 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUQGsYUk48loUBwANY:-KqUQGsYUk48loUBwANZ:b-parr0z) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L327)):*
> ```Go
> 	}
> 
> 	tryLogin := func(username, password string) (http.Header, error) {
> ```

Maybe it's just me but I don't like variable shadowing.

---

*[pkg/server/serverpb/authentication.proto, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqU7rC6DosxXR6oX5KJ:-KqU7rC7YU3NmcPn0Rrq:bm23pka) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/serverpb/authentication.proto#L76)):*
> ```protobuf
>     };
>   }
> }
> ```

Trailing newline?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-:-KqUSB-8_6TgD2Uyhr4N:b-usiczd)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 16 files at r1.
Review status: 8 of 16 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[pkg/server/authentication.go, line 54 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqXdB7GvAXgt5EW4Dfd:-KqXdB7GvAXgt5EW4Dfe:bf7u98r) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L54)):*
> ```Go
> var errBadCredentials = grpc.Errorf(
> 	codes.Unauthenticated,
> 	"The provided username and password did not match any credentials on the server.",
> ```

i believe errors should neither be capitalized nor have trailing punctuation (there are a few instances in this commit)

---

*[pkg/server/authentication_test.go, line 209 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqXeQ3qCTZs2-Bh81dX:-KqXeQ3qCTZs2-Bh81dY:b-atles8) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication_test.go#L209)):*
> ```Go
> 				}
> 			} else {
> 				if err == nil {
> ```

don't you want to check for the specific error you expect?

---

*[pkg/server/serverpb/authentication.proto, line 52 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqXejhxG-OHIj8uNzqO:-KqXejhxG-OHIj8uNzqP:biov06d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/serverpb/authentication.proto#L52)):*
> ```protobuf
> // from successful login requests.
> message SessionCookie {
>   // The unique ID of the session.
> ```

inconsistent indentation

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 14 of 16 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 20 unresolved discussions, some commit checks failed.

---

*[pkg/server/authentication.go, line 109 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZS-Y2hEO3HFbwIRMk:-KqZS-Y2hEO3HFbwIRMl:bxd807i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L109)):*
> ```Go
> 	username := req.Username
> 	if username == "" {
> 		return nil, errBadCredentials
> ```

Optional, but it'd be nice to be more specific and say that a username wasn't provided.

---

*[pkg/server/authentication.go, line 114 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZSCgB5d5qMhYkgJXP:-KqZSCgB5d5qMhYkgJXQ:bap4uun) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L114)):*
> ```Go
> 	// Root user does not have a password, simply disallow this.
> 	if username == security.RootUser {
> 		return nil, errBadCredentials
> ```

Ditto, this error could be made more helpful.

---

*[pkg/server/authentication.go, line 171 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZSi4TB1fy4nN0aYer:-KqZSi4TB1fy4nN0aYes:byj15lp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L171)):*
> ```Go
> 	)
> 	if err != nil {
> 		return err
> ```

It'd be nice to distinguish between internal errors and bad-credentials errors.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-:-KqZVNyW4H2NFOoQw9Lp:bnfp4nl)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 7 of 16 files reviewed at latest revision, 20 unresolved discussions.

---

*[pkg/server/api_error.go, line 15 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTi4m8-sL4yUNEgOJK:-KqZ31sQ3PSXS3In5SxU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/api_error.go#L15)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

FYI since #17346 Author comments are no bueno.
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 59 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTkVfeDJLuriD0eWEA:-KqZ3KdbU6BVS_JgyajL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L59)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Nits:
- Looks like these descriptions don't usually have trailing punctuation.
- I'd remove "login".
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 65 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTm6nPxT9WfSBZaGq0:-KqZ3J9Hg0PZ2B48Ijhz:b-8t1mtz) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L65)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

It doesn't look like this is used anywhere?
</blockquote></details>

Yeah, cargo culted. Removed.

---

*[pkg/server/authentication.go, line 103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTn5UM7wlELZ0ERo7R:-KqZ3ctKjugtuF475mBu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L103)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Looks like this comment is unfinished?
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 163 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqTpbm-vDzPpPvxvDpB:-KqZ3mmB067vlBwNgrps:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L163)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

To make everything super clear, it might be useful to repeat the comment you added to `CompareHashAndPassword`, specifically "If they are not equivalent, returns an error."
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqU5KRYaBHcx_n22Pi9:-KqZ3xS6QAmbK6mud_No:bd7hvpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L220)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Maybe it's just me, but I find it really hard to read a multi-line `if` initialization.
</blockquote></details>

I believe this style is actually linter-enforced.

---

*[pkg/server/authentication.go, line 229 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqU5jqLZRMMIcJJTp98:-KqZ4DT1LpSBpoAMo8E3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L229)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Looks like this comment is unfinished.
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 54 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqXdB7GvAXgt5EW4Dfd:-KqZ3KpCAMU_rH6DYfuE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L54)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i believe errors should neither be capitalized nor have trailing punctuation (there are a few instances in this commit)
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 109 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZS-Y2hEO3HFbwIRMk:-KqZ_XmXxcfB5qNDWg3u:b-nxf9dp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L109)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Optional, but it'd be nice to be more specific and say that a username wasn't provided.
</blockquote></details>

In this case, I think it's likely fine to provide a more specific error message, since it's a request that is bad in a specific way...

---

*[pkg/server/authentication.go, line 114 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZSCgB5d5qMhYkgJXP:-KqZ_h7jmq7RbuACOZcK:b-qqtwc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L114)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Ditto, this error could be made more helpful.
</blockquote></details>

However, when actual credentials are involved, I think it's best practice to be deliberately non-specific in order to avoid providing an oracle.

---

*[pkg/server/authentication.go, line 171 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZSi4TB1fy4nN0aYer:-KqZdyknsod0HQ1aWpke:b-5qvibd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L171)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It'd be nice to distinguish between internal errors and bad-credentials errors.
</blockquote></details>

There is some work involved in this which must be pushed down to GetUserHashedPassword, so I'm going to mark it as a TODO.

---

*[pkg/server/authentication_test.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUApxBVlPerZO-4Z4m:-KqZ4K4DI6lXjncl167c:bdq52ad) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L185)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I'd move these test cases out above the `for` statement.
</blockquote></details>

This is a "table-driven test", a common Go style. This factoring is suggested directly from the golang project itself: https://github.com/golang/go/wiki/TableDrivenTests

---

*[pkg/server/authentication_test.go, line 191 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUNBAROXQOthArSHWv:-KqZ59Xv4yBG4aa-Ye1B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L191)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Just for good measure, how about a suffix match like `shunter2`?
</blockquote></details>

Done.

---

*[pkg/server/authentication_test.go, line 270 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUOOmB5zrM3QUIUvx-:-KqZYJ9Pv14EtYr4-qBh:bp3qchs) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L270)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Should this error also say the word "secret" somewhere?  Also, spitting out the raw bytes might not be a good idea (how does `%v` for `[]byte` look?)
</blockquote></details>

The alternatives would be %s, which is equally inscrutable, or `%#v` which prints out go syntax to create the same byte array.  None of them are human readable, but I think they have value since they place the exact values in the error output.

I've modified this to use `%#v` and some newlines, now has the form:

```
hashed value of secret: 
		[]byte{0x49, 0xc7, 0xae, 0xd3, 0x36, 0x27, 0x36, 0x47, 0xc5, 0xc2, 0xb0, 0x6f, 0x67, 0xa3, 0xd3, 0x68}
		computed as: 
		[]byte{0x49, 0xc7, 0xae, 0xd3, 0x36, 0x27, 0x36, 0x47, 0xc5, 0xc2, 0xb0, 0x6f, 0x67, 0xa3, 0xd3, 0x68, 0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14, 0x9a, 0xfb, 0xf4, 0xc8, 0x99, 0x6f, 0xb9, 0x24, 0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b, 0x93, 0x4c, 0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55}
		wanted: 
		[]byte{0x49, 0xc7, 0xae, 0xd3, 0x36, 0x27, 0x36, 0x47, 0xc5, 0xc2, 0xb0, 0x6f, 0x67, 0xa3, 0xd3, 0x68, 0xe3, 0xb0, 0xc4, 0x42, 0x98, 0xfc, 0x1c, 0x14, 0x9a, 0xfb, 0xf4, 0xc8, 0x99, 0x6f, 0xb9, 0x24, 0x27, 0xae, 0x41, 0xe4, 0x64, 0x9b, 0x93, 0x4c, 0xa4, 0x95, 0x99, 0x1b, 0x78, 0x52, 0xb8, 0x55}
```

---

*[pkg/server/authentication_test.go, line 274 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUOoHx37UVHlSrPMyr:-KqZVwmuz0t0AJEpwFGY:b-5491u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L274)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Why the aliasing?
</blockquote></details>

The pattern works really well when you compare nested arguments (e.g. `if a, e := myResponse.SubObject.anotherObject, previouslyComputedExpectedObject; a != e`).   The benefit is that you can use `a` and `e` inside of the if block when constructing error messages, meaning you don't have to repeat the extraction.

It's not strictly necessary here, but I still use this pattern in most places.

---

*[pkg/server/authentication_test.go, line 283 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUP6mjtiVO7JO_KhbG:-KqZYGBu9Hlo9DnFW3kQ:b-l2tt7p) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L283)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

:s/early/actual/
</blockquote></details>

Done, good catch.

---

*[pkg/server/authentication_test.go, line 327 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUQGsYUk48loUBwANY:-KqZX4MuYieJyNprsYVU:buegovc) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L327)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Maybe it's just me but I don't like variable shadowing.
</blockquote></details>

It's not just you, this kind of thing is usually linter enforced, suppose it doesn't apply to tests.  Changed.

---

*[pkg/server/authentication_test.go, line 209 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqXeQ3qCTZs2-Bh81dX:-KqZVpVvCZ1knWxUP4cW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication_test.go#L209)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

don't you want to check for the specific error you expect?
</blockquote></details>

Done.

---

*[pkg/server/serverpb/authentication.proto, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqU7rC6DosxXR6oX5KJ:-KqZebV_zc-ghONsUguB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/serverpb/authentication.proto#L76)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Trailing newline?
</blockquote></details>

Done.

---

*[pkg/server/serverpb/authentication.proto, line 52 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqXejhxG-OHIj8uNzqO:-KqZYCblipc8ah-K-pgd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/serverpb/authentication.proto#L52)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

inconsistent indentation
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 9 files at r3.
Review status: 14 of 16 files reviewed at latest revision, 18 unresolved discussions.

---

*[pkg/server/authentication.go, line 23 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZiwlilKMerjDefA1_:-KqZiwlilKMerjDefA1a:b-1nzhwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b37a7946e619247b1705ec052915c5e5a495bfc/pkg/server/authentication.go#L23)):*
> ```Go
> 	"time"
> 
> 	"fmt"
> ```

can you group all the stdlib imports? while you're in there, you could fix the other grouping, but that's optional.

there are a few instances of oddly grouped standard imports.

---

*[pkg/server/authentication.go, line 75 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZj3M9PwmSgCUy83fl:-KqZj3M9PwmSgCUy83fm:b-553mla) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b37a7946e619247b1705ec052915c5e5a495bfc/pkg/server/authentication.go#L75)):*
> ```Go
> 		memMetrics: &s.adminMemMetrics,
> 	}
> 	return server
> ```

no need for the local - also, no need to use a pointer here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343)*
<!-- Sent from Reviewable.io -->

couchand commented:
Review status: 14 of 16 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/server/authentication.go, line 220 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqU5KRYaBHcx_n22Pi9:-KqZkwCO3p_qKF21O5e5:bfgpsg0) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication.go#L220)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

I believe this style is actually linter-enforced.
</blockquote></details>

Fair enough.  With a couple lines it's not too bad, but when it's super long like this it's hard to track the `if` keyword with the block that follows.  I may just go hunting for that lint rule...

---

*[pkg/server/authentication.go, line 114 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZSCgB5d5qMhYkgJXP:-KqZmMJnYR8EWhDsPT-P:b-dk8qyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L114)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

However, when actual credentials are involved, I think it's best practice to be deliberately non-specific in order to avoid providing an oracle.
</blockquote></details>

Definitely agree that not providing details here is the right way to go.

---

*[pkg/server/authentication_test.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUApxBVlPerZO-4Z4m:-KqZlK9TYn-t1Z8qh0YU:b-ugsnr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L185)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

This is a "table-driven test", a common Go style. This factoring is suggested directly from the golang project itself: https://github.com/golang/go/wiki/TableDrivenTests
</blockquote></details>

I'm all about the table tests.  I just meant that the test cases could be defined above the `for` statement.  Same general principle as my multi-line `if` initializer complaint.  Seems like it's inconsistent, but at least a few other tests define the slice of test cases in a freestanding definition statement rather than within the `for` initializer.

---

*[pkg/server/authentication_test.go, line 274 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUOoHx37UVHlSrPMyr:-KqZm-GChy00r6NmlEos:b-vc1tz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L274)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

The pattern works really well when you compare nested arguments (e.g. `if a, e := myResponse.SubObject.anotherObject, previouslyComputedExpectedObject; a != e`).   The benefit is that you can use `a` and `e` inside of the if block when constructing error messages, meaning you don't have to repeat the extraction.

It's not strictly necessary here, but I still use this pattern in most places.
</blockquote></details>

Looks really useful for deep access, seems reasonable to just get in the habit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 13 of 18 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/server/authentication.go, line 114 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZSCgB5d5qMhYkgJXP:-KqZqw7NYwh5abC1Hqxa:b-9dvt0s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L114)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

Definitely agree that not providing details here is the right way to go.
</blockquote></details>

I'm going to use the error message from PGWire; after thinking about, the username of our RootUser is hard coded, so I don't think this leaks any information.

---

*[pkg/server/authentication.go, line 171 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZSi4TB1fy4nN0aYer:-KqZrBFvA2auJj7iOPST:b9ev5qr) ([raw file](https://github.com/cockroachdb/cockroach/blob/a62dd76e706a6b95714021faea3b0bf07d660b3b/pkg/server/authentication.go#L171)):*
<details><summary><i>Previously, mrtracy (Matt Tracy) wrote…</i></summary><blockquote>

There is some work involved in this which must be pushed down to GetUserHashedPassword, so I'm going to mark it as a TODO.
</blockquote></details>

Actually, I ran into an issue where this would be helpful on my next PR. I'm taking this change, internal server errors are now distinguished from "user doesn't exist".

---

*[pkg/server/authentication.go, line 23 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZiwlilKMerjDefA1_:-KqZk0sujaacUBuYZmQG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b37a7946e619247b1705ec052915c5e5a495bfc/pkg/server/authentication.go#L23)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you group all the stdlib imports? while you're in there, you could fix the other grouping, but that's optional.

there are a few instances of oddly grouped standard imports.
</blockquote></details>

Done.

---

*[pkg/server/authentication.go, line 75 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqZj3M9PwmSgCUy83fl:-KqZk44Os5pAUIp8Zwf_:b-e827y3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b37a7946e619247b1705ec052915c5e5a495bfc/pkg/server/authentication.go#L75)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

no need for the local - also, no need to use a pointer here.
</blockquote></details>

Removed the local, but I think the pointer should be kept just for consistency with the other server objects, which are instantiated as pointers using "new" methods.

---

*[pkg/server/authentication_test.go, line 185 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343#-KqUApxBVlPerZO-4Z4m:-KqZrvytye8kfig2p189:b-d8l86g) ([raw file](https://github.com/cockroachdb/cockroach/blob/32964c2cbdab9a0c1b4572581b6eb6152b70b664/pkg/server/authentication_test.go#L185)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I'm all about the table tests.  I just meant that the test cases could be defined above the `for` statement.  Same general principle as my multi-line `if` initializer complaint.  Seems like it's inconsistent, but at least a few other tests define the slice of test cases in a freestanding definition statement rather than within the `for` initializer.
</blockquote></details>

I've personally done it this way almost everywhere.  The "for" is essentially elided in my mind, when I see a block like this I know exactly what's going on.  I don't think there is sufficiently complicated information present in the "for" statement that it needs to be moved closer to the function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17343)*
<!-- Sent from Reviewable.io -->
